### ISSUE TITLE
Creating CodeBuildCredentials via Groovy?

### ISSUE THREAD

**Issue by robbinscpcg** [awslabs/aws-codebuild-jenkins-plugin#79]
I'm trying to create CodeBuild credentials using Groovy as a part of our Jenkins automation.  Is there a documented constructor, or am I missing a plugin?  

Sample Code

````
import jenkins.model.*
import com.cloudbees.plugins.credentials.*
import com.cloudbees.plugins.credentials.common.*
import com.cloudbees.plugins.credentials.domains.*
import com.cloudbees.jenkins.plugins.awscredentials.*
import com.cloudbees.jenkins.plugins.sshcredentials.impl.*
import com.dabsquared.gitlabjenkins.connection.GitLabApiTokenImpl
import org.jenkinsci.plugins.plaincredentials.*
import org.jenkinsci.plugins.plaincredentials.impl.*
import hudson.util.Secret
import com.cloudbees.plugins.credentials.impl.*;
import com.cloudbees.plugins.credentials.*;
import com.cloudbees.plugins.credentials.domains.*;
import com.amazonaws.services.codebuild.AWSCodeBuildClient;
import com.amazonaws.services.codebuild.model.*;
import com.amazonaws.services.codebuild.model.Build;
import com.cloudbees.hudson.plugins.folder.Folder;
import com.cloudbees.plugins.credentials.Credentials;
import com.cloudbees.plugins.credentials.CredentialsProvider;

domain = Domain.global()
store = Jenkins.instance.getExtensionList('com.cloudbees.plugins.credentials.SystemCredentialsProvider')[0].getStore()


awsCodeBuild = new CodeBuildCredentials(
    CredentialsScope.GLOBAL, 
    "awsCodeBuild", 
    "AWS CodeBuild Credentials", 
    Secret.fromString("AWS_ACCESS_KEY_ID"), 
    Secret.fromString("AWS_SECRET_ACCESS_KEY"),
    "", 
    "", 
    "", 
    ""
) 

store.addCredentials(domain, awsCodeBuild)
````

The code above throws the following error.  

````
groovy.lang.GroovyRuntimeException: Could not find matching constructor for: CodeBuildCredentials(com.cloudbees.plugins.credentials.CredentialsScope$2, java.lang.String, java.lang.String, hudson.util.Secret, hudson.util.Secret, java.lang.String, java.lang.String, java.lang.String, java.lang.String)
	at groovy.lang.MetaClassImpl.invokeConstructor(MetaClassImpl.java:1732)
	at groovy.lang.MetaClassImpl.invokeConstructor(MetaClassImpl.java:1532)
	at org.codehaus.groovy.runtime.callsite.MetaClassConstructorSite.callConstructor(MetaClassConstructorSite.java:49)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:60)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:235)
	at Script1.run(Script1.groovy:33)
````

**Comment by leobaran-aws** [comment]
I'm not actually sure if CodeBuildCredentials supports this at the moment. I'll look into it and get back to you.

**Comment by leobaran-aws** [comment]
I think what we need to do is model CodeBuildCredentials in a java package. Then you would be able to import it with something like `import com.amazonaws.codebuild.jenkinsplugin.CodeBuildCredentials;`. This would probably be a backward-incompatible change--I'm looking into how much it will affect existing projects and credentials. Worst case, we'll do a major version update and Jenkins users will have to recreate their non-pipeline projects and credentials.

**Comment by robbinscpcg** [comment]
@leobaran-aws - This would be **greatly** appreciated.  I know from our perspective, it'd cut down on a lot of manual work post-deploy on Jenkins.

**Comment by cocooma** [comment]
Hi guys,

Have you guys made any progress in regards the above?

**Comment by cocooma** [comment]
May I ask what is the reason behind that the code lives in the default package space instead of having it structured/separated  with packages?

**Comment by leoherran-aws** [comment]
Putting the plugin code in the default package space was an unfortunate design flaw. We plan to address this with a major release sometime in the future. 

I implemented a fix that adds a wrapper around `CodeBuildCredentials` which will give you access to a groovy-compatible credentials class. It should be pushed and released within the next couple of days--thanks for your patience.

**Comment by leoherran-aws** [comment]
Implemented in d45f586761c3f479f0b0ad2ada2f80801d5f1045. You'll need to add the following import statement to your scripts: 

`import com.amazonaws.codebuild.jenkinsplugin.CodeBuildBaseCredentials;`

And then you can instantiate credentials like so: `awsCodeBuild = new CodeBuildBaseCredentials(...)`

This will be released in version `0.45` which should be available later today. Let us know if it works for your use case.
